File: dream/cron.py
"""A small five-field cron parser, matcher and next-fire calculator.

Dream schedules are stored as standard cron expressions because that is the
format users can read, edit and paste elsewhere. Only the classic five fields
are supported — minute, hour, day-of-month, month, day-of-week — with ``*``,
``*/n``, ``a-b``, ``a-b/n`` and comma lists. No seconds field, no ``@hourly``
aliases, no ``L``/``W``/``#`` extensions: the scheduler polls every thirty
seconds, so sub-minute precision would be a promise it cannot keep.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta

__all__ = [
    "CRON_FIELDS",
    "CronExpression",
    "cron_matches",
    "describe_cron",
    "next_run_after",
    "parse_cron",
    "validate_cron",
]

#: ``(name, minimum, maximum)`` per field, in wire order.
CRON_FIELDS: tuple[tuple[str, int, int], ...] = (
    ("minute", 0, 59),
    ("hour", 0, 23),
    ("day", 1, 31),
    ("month", 1, 12),
    ("weekday", 0, 6),
)

_MONTH_NAMES = (
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
)
_WEEKDAY_NAMES = (
    "Sunday",
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
)

#: Four years of minutes is the bound on the forward walk in
#: :func:`next_run_after`. It comfortably covers the longest real schedule (29
#: February, which recurs at most every four years) while guaranteeing that an
#: impossible expression such as ``0 0 30 2 *`` terminates.
_SEARCH_LIMIT_DAYS = 366 * 4


@dataclass(frozen=True, slots=True)
class CronExpression:
    """A parsed cron expression as five sets of permitted values."""

    expression: str
    minutes: frozenset[int]
    hours: frozenset[int]
    days: frozenset[int]
    months: frozenset[int]
    weekdays: frozenset[int]
    day_restricted: bool
    weekday_restricted: bool

    def matches(self, moment: datetime) -> bool:
        """Whether ``moment`` (to the minute) satisfies this expression."""
        if moment.minute not in self.minutes or moment.hour not in self.hours:
            return False
        if moment.month not in self.months:
            return False
        # Vixie cron: when both day-of-month and day-of-week are restricted the
        # expression fires on either, not on their intersection. "1st of the
        # month and every Monday" must not silently mean "Monday the 1st".
        day_ok = moment.day in self.days
        weekday_ok = ((moment.weekday() + 1) % 7) in self.weekdays
        if self.day_restricted and self.weekday_restricted:
            return day_ok or weekday_ok
        return day_ok and weekday_ok


def _parse_field(raw: str, name: str, low: int, high: int) -> tuple[frozenset[int], bool]:
    """Expand one field to its value set and whether it is restricted."""
    raw = raw.strip()
    if not raw:
        raise ValueError(f"cron {name} field is empty")
    values: set[int] = set()
    restricted = False
    for part in raw.split(","):
        part = part.strip()
        if not part:
            raise ValueError(f"cron {name} field has an empty list item")
        step = 1
        if "/" in part:
            part, _, step_text = part.partition("/")
            try:
                step = int(step_text)
            except ValueError:
                raise ValueError(
                    f"cron {name} step must be an integer, got {step_text!r}"
                ) from None
            if step < 1:
                raise ValueError(f"cron {name} step must be positive, got {step}")
        if part in ("*", ""):
            start, end = low, high
            if step != 1:
                restricted = True
        elif "-" in part.lstrip("-"):
            start_text, _, end_text = part.partition("-")
            start, end = _int(start_text, name), _int(end_text, name)
            restricted = True
        else:
            start = end = _int(part, name)
            restricted = True
        if name == "weekday":
            # Both 0 and 7 mean Sunday in every cron dialect worth matching.
            start, end = (0 if start == 7 else start), (0 if end == 7 else end)
            if end < start:
                # A wrapping range such as fri-mon (5-1).
                values.update(range(start, 7, step))
                values.update(range(0, end + 1, step))
                continue
        if start < low or end > high or end < start:
            raise ValueError(f"cron {name} value out of range [{low}, {high}]: {part!r}")
        values.update(range(start, end + 1, step))
    if not values:
        raise ValueError(f"cron {name} field matches nothing: {raw!r}")
    return frozenset(values), restricted


def _int(text: str, field: str) -> int:
    try:
        return int(text.strip())
    except ValueError:
        raise ValueError(f"cron {field} value must be an integer, got {text.strip()!r}") from None


def parse_cron(expression: str) -> CronExpression:
    """Parse a five-field cron expression, raising ``ValueError`` if invalid."""
    if not isinstance(expression, str):
        raise ValueError("cron expression must be a string")
    fields = expression.split()
    if len(fields) != 5:
        raise ValueError(
            f"cron expression must have 5 fields (minute hour day month weekday), "
            f"got {len(fields)}: {expression!r}"
        )
    parsed: list[frozenset[int]] = []
    restrictions: list[bool] = []
    for raw, (name, low, high) in zip(fields, CRON_FIELDS, strict=True):
        values, restricted = _parse_field(raw, name, low, high)
        parsed.append(values)
        restrictions.append(restricted)
    return CronExpression(
        expression=" ".join(fields),
        minutes=parsed[0],
        hours=parsed[1],
        days=parsed[2],
        months=parsed[3],
        weekdays=parsed[4],
        day_restricted=restrictions[2],
        weekday_restricted=restrictions[4],
    )


def validate_cron(expression: str) -> str:
    """Return the normalised expression, or raise ``ValueError``."""
    return parse_cron(expression).expression


def cron_matches(expression: str | CronExpression, moment: datetime) -> bool:
    """Whether a cron expression fires at ``moment``."""
    parsed = expression if isinstance(expression, CronExpression) else parse_cron(expression)
    return parsed.matches(moment)


def next_run_after(expression: str | CronExpression, after: datetime) -> datetime:
    """The first minute strictly after ``after`` at which the expression fires.

    Walks forward a minute at a time, skipping whole days when the date fields
    cannot match, which keeps even a yearly schedule well under a millisecond.
    """
    parsed = expression if isinstance(expression, CronExpression) else parse_cron(expression)
    moment = (after + timedelta(minutes=1)).replace(second=0, microsecond=0)
    limit = after + timedelta(days=_SEARCH_LIMIT_DAYS)
    while moment <= limit:
        if not _date_could_match(parsed, moment):
            moment = (moment + timedelta(days=1)).replace(hour=0, minute=0)
            continue
        if parsed.matches(moment):
            return moment
        moment += timedelta(minutes=1)
    raise ValueError(f"cron expression never fires: {parsed.expression!r}")


def _date_could_match(parsed: CronExpression, moment: datetime) -> bool:
    if moment.month not in parsed.months:
        return False
    day_ok = moment.day in parsed.days
    weekday_ok = ((moment.weekday() + 1) % 7) in parsed.weekdays
    if parsed.day_restricted and parsed.weekday_restricted:
        return day_ok or weekday_ok
    return day_ok and weekday_ok


# --------------------------------------------------------------------------
# Human-readable rendering
# --------------------------------------------------------------------------


def _ordinal(n: int) -> str:
    if 11 <= n % 100 <= 13:
        return f"{n}th"
    return f"{n}{ {1: 'st', 2: 'nd', 3: 'rd'}.get(n % 10, 'th') }"


def _clock(hour: int, minute: int) -> str:
    suffix = "AM" if hour < 12 else "PM"
    display = hour % 12 or 12
    return f"{display}:{minute:02d} {suffix}"


def _step_of(field: str) -> int | None:
    if field.startswith("*/"):
        try:
            return int(field[2:])
        except ValueError:
            return None
    return None


def _weekday_phrase(field: str, weekdays: frozenset[int]) -> str:
    if field == "1-5":
        return "every weekday"
    if field in ("0,6", "6,0", "6,7", "0,6,7"):
        return "every weekend day"
    if field == "6,0,1,2,3":
        return "every Iranian working day (Sat–Wed)"
    names = [_WEEKDAY_NAMES[d] for d in sorted(weekdays)]
    if len(names) == 1:
        return f"every {names[0]}"
    return "every " + ", ".join(names[:-1]) + f" and {names[-1]}"


def describe_cron(expression: str) -> str:
    """Render a cron expression as an English sentence for the UI.

    Covers the shapes :func:`dream.scheduler.nl_to_cron` can emit precisely,
    and degrades to a field-by-field reading for anything hand-written.
    """
    parsed = parse_cron(expression)
    minute_f, hour_f, day_f, month_f, weekday_f = parsed.expression.split()

    minute_step = _step_of(minute_f)
    if minute_step and hour_f == "*" and day_f == "*" and month_f == "*" and weekday_f == "*":
        unit = "minute" if minute_step == 1 else "minutes"
        return f"every {minute_step} {unit}"

    hour_step = _step_of(hour_f)
    if hour_step and day_f == "*" and month_f == "*" and weekday_f == "*":
        minutes = sorted(parsed.minutes)
        at = "" if minutes == [0] else f" at {minutes[0]} past"
        unit = "hour" if hour_step == 1 else "hours"
        return f"every {hour_step} {unit}{at}"

    day_step = _step_of(day_f)
    if day_step and month_f == "*" and weekday_f == "*" and len(parsed.hours) == 1:
        hour = next(iter(parsed.hours))
        minute = min(parsed.minutes)
        unit = "day" if day_step == 1 else "days"
        return f"every {day_step} {unit} at {_clock(hour, minute)}"

    if minute_f == "*" and hour_f == "*":
        return "every minute"

    if len(parsed.hours) != 1 or len(parsed.minutes) != 1:
        return f"at cron schedule {parsed.expression}"

    at = _clock(next(iter(parsed.hours)), next(iter(parsed.minutes)))

    if weekday_f != "*":
        when = _weekday_phrase(weekday_f, parsed.weekdays)
    elif day_f != "*":
        days = " and ".join(_ordinal(d) for d in sorted(parsed.days))
        when = f"on the {days}"
    else:
        when = "every day"

    if month_f != "*":
        months = " and ".join(_MONTH_NAMES[m - 1] for m in sorted(parsed.months))
        when = f"{when} in {months}"

    return f"{when} at {at}"

File: apps/desktop/src/lib/schedule/cron.ts
/**
 * Cron parsing, matching and human-readable rendering.
 *
 * A faithful TypeScript mirror of `dream/cron.py`: five fields only
 * (star, step, range and comma lists), no seconds, no `@aliases`, no
 * `L`/`W`/`#` extensions.
 * The Python sidecar remains the source of truth for schedules that actually
 * fire; this port exists so the UI can validate input, render a live preview
 * and show the next fire time without a round trip on every keystroke.
 */

/** `[name, minimum, maximum]` per field, in wire order. */
export const CRON_FIELDS = [
  ['minute', 0, 59],
  ['hour', 0, 23],
  ['day', 1, 31],
  ['month', 1, 12],
  ['weekday', 0, 6],
] as const;

const MONTH_NAMES = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
] as const;

const WEEKDAY_NAMES = [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
] as const;

/**
 * Four years of days bounds the forward walk in {@link nextRunAfter}: enough
 * for 29 February, finite for an impossible expression such as `0 0 30 2 *`.
 */
const SEARCH_LIMIT_DAYS = 366 * 4;

/** A parsed cron expression as five sets of permitted values. */
export interface CronExpression {
  expression: string;
  minutes: ReadonlySet<number>;
  hours: ReadonlySet<number>;
  days: ReadonlySet<number>;
  months: ReadonlySet<number>;
  weekdays: ReadonlySet<number>;
  dayRestricted: boolean;
  weekdayRestricted: boolean;
}

/** Raised for any malformed expression. */
export class CronError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CronError';
  }
}

function toInt(text: string, field: string): number {
  const trimmed = text.trim();
  if (!/^-?\d+$/.test(trimmed)) {
    throw new CronError(`cron ${field} value must be an integer, got '${trimmed}'`);
  }
  return Number.parseInt(trimmed, 10);
}

/** Expands one field to its value set and whether it is restricted. */
function parseField(raw: string, name: string, low: number, high: number): [Set<number>, boolean] {
  const field = raw.trim();
  if (!field) throw new CronError(`cron ${name} field is empty`);

  const values = new Set<number>();
  let restricted = false;

  for (const item of field.split(',')) {
    let part = item.trim();
    if (!part) throw new CronError(`cron ${name} field has an empty list item`);

    let step = 1;
    const slash = part.indexOf('/');
    if (slash !== -1) {
      const stepText = part.slice(slash + 1);
      part = part.slice(0, slash);
      if (!/^\d+$/.test(stepText.trim())) {
        throw new CronError(`cron ${name} step must be an integer, got '${stepText}'`);
      }
      step = Number.parseInt(stepText, 10);
      if (step < 1) throw new CronError(`cron ${name} step must be positive, got ${step}`);
    }

    let start: number;
    let end: number;
    if (part === '*' || part === '') {
      start = low;
      end = high;
      if (step !== 1) restricted = true;
    } else if (part.replace(/^-+/, '').includes('-')) {
      const dash = part.indexOf('-');
      start = toInt(part.slice(0, dash), name);
      end = toInt(part.slice(dash + 1), name);
      restricted = true;
    } else {
      start = end = toInt(part, name);
      restricted = true;
    }

    if (name === 'weekday') {
      // Both 0 and 7 mean Sunday in every cron dialect worth matching.
      if (start === 7) start = 0;
      if (end === 7) end = 0;
      if (end < start) {
        // A wrapping range such as fri-mon (5-1).
        for (let v = start; v <= 6; v += step) values.add(v);
        for (let v = 0; v <= end; v += step) values.add(v);
        continue;
      }
    }

    if (start < low || end > high || end < start) {
      throw new CronError(`cron ${name} value out of range [${low}, ${high}]: '${part}'`);
    }
    for (let v = start; v <= end; v += step) values.add(v);
  }

  if (values.size === 0) throw new CronError(`cron ${name} field matches nothing: '${field}'`);
  return [values, restricted];
}

/** Parses a five-field cron expression, throwing {@link CronError} if invalid. */
export function parseCron(expression: string): CronExpression {
  if (typeof expression !== 'string') throw new CronError('cron expression must be a string');
  const fields = expression.trim().split(/\s+/).filter(Boolean);
  if (fields.length !== 5) {
    throw new CronError(
      `cron expression must have 5 fields (minute hour day month weekday), ` +
        `got ${fields.length}: '${expression}'`,
    );
  }
  const sets: Set<number>[] = [];
  const restrictions: boolean[] = [];
  fields.forEach((raw, index) => {
    const [name, low, high] = CRON_FIELDS[index];
    const [values, restricted] = parseField(raw, name, low, high);
    sets.push(values);
    restrictions.push(restricted);
  });
  return {
    expression: fields.join(' '),
    minutes: sets[0],
    hours: sets[1],
    days: sets[2],
    months: sets[3],
    weekdays: sets[4],
    dayRestricted: restrictions[2],
    weekdayRestricted: restrictions[4],
  };
}

/** Returns the normalised expression, or throws. */
export function validateCron(expression: string): string {
  return parseCron(expression).expression;
}

/** Whether a cron expression fires at `moment` (to the minute). */
export function cronMatches(expression: string | CronExpression, moment: Date): boolean {
  const parsed = typeof expression === 'string' ? parseCron(expression) : expression;
  if (!parsed.minutes.has(moment.getMinutes())) return false;
  if (!parsed.hours.has(moment.getHours())) return false;
  if (!parsed.months.has(moment.getMonth() + 1)) return false;
  // Vixie cron: when both day-of-month and day-of-week are restricted the
  // expression fires on either, not on their intersection.
  const dayOk = parsed.days.has(moment.getDate());
  const weekdayOk = parsed.weekdays.has(moment.getDay());
  if (parsed.dayRestricted && parsed.weekdayRestricted) return dayOk || weekdayOk;
  return dayOk && weekdayOk;
}

function dateCouldMatch(parsed: CronExpression, moment: Date): boolean {
  if (!parsed.months.has(moment.getMonth() + 1)) return false;
  const dayOk = parsed.days.has(moment.getDate());
  const weekdayOk = parsed.weekdays.has(moment.getDay());
  if (parsed.dayRestricted && parsed.weekdayRestricted) return dayOk || weekdayOk;
  return dayOk && weekdayOk;
}

/**
 * The first minute strictly after `after` at which the expression fires.
 *
 * Walks forward a minute at a time, skipping whole days whose date fields
 * cannot match, which keeps even a yearly schedule well under a millisecond.
 */
export function nextRunAfter(expression: string | CronExpression, after: Date): Date {
  const parsed = typeof expression === 'string' ? parseCron(expression) : expression;
  const moment = new Date(after.getTime() + 60_000);
  moment.setSeconds(0, 0);
  const limit = after.getTime() + SEARCH_LIMIT_DAYS * 86_400_000;

  while (moment.getTime() <= limit) {
    if (!dateCouldMatch(parsed, moment)) {
      moment.setDate(moment.getDate() + 1);
      moment.setHours(0, 0, 0, 0);
      continue;
    }
    if (cronMatches(parsed, moment)) return moment;
    moment.setTime(moment.getTime() + 60_000);
  }
  throw new CronError(`cron expression never fires: '${parsed.expression}'`);
}

/** The next `count` fire times after `after`. */
export function upcomingRuns(
  expression: string | CronExpression,
  count = 3,
  after: Date = new Date(),
): Date[] {
  const parsed = typeof expression === 'string' ? parseCron(expression) : expression;
  const out: Date[] = [];
  let cursor = after;
  for (let i = 0; i < count; i += 1) {
    cursor = nextRunAfter(parsed, cursor);
    out.push(cursor);
  }
  return out;
}

// --------------------------------------------------------------------------- //
// Human-readable rendering
// --------------------------------------------------------------------------- //

function ordinal(n: number): string {
  if (n % 100 >= 11 && n % 100 <= 13) return `${n}th`;
  return `${n}${({ 1: 'st', 2: 'nd', 3: 'rd' } as Record<number, string>)[n % 10] ?? 'th'}`;
}

/** Renders `13:05` as `1:05 PM`. */
export function clockLabel(hour: number, minute: number): string {
  const suffix = hour < 12 ? 'AM' : 'PM';
  const display = hour % 12 || 12;
  return `${display}:${String(minute).padStart(2, '0')} ${suffix}`;
}

function stepOf(field: string): number | null {
  if (!field.startsWith('*/')) return null;
  const step = Number.parseInt(field.slice(2), 10);
  return Number.isNaN(step) ? null : step;
}

function weekdayPhrase(field: string, weekdays: ReadonlySet<number>): string {
  if (field === '1-5') return 'every weekday';
  if (['0,6', '6,0', '6,7', '0,6,7'].includes(field)) return 'every weekend day';
  if (field === '6,0,1,2,3') return 'every Iranian working day (Sat–Wed)';
  const names = [...weekdays].sort((a, b) => a - b).map((d) => WEEKDAY_NAMES[d]);
  if (names.length === 1) return `every ${names[0]}`;
  return `every ${names.slice(0, -1).join(', ')} and ${names[names.length - 1]}`;
}

/**
 * Renders a cron expression as an English sentence for the UI.
 *
 * Covers every shape {@link nlToCron} can emit precisely, and degrades to a
 * literal reading for anything hand-written.
 */
export function describeCron(expression: string): string {
  const parsed = parseCron(expression);
  const [minuteF, hourF, dayF, monthF, weekdayF] = parsed.expression.split(' ') as [
    string,
    string,
    string,
    string,
    string,
  ];

  const minuteStep = stepOf(minuteF);
  if (minuteStep && hourF === '*' && dayF === '*' && monthF === '*' && weekdayF === '*') {
    return `every ${minuteStep} ${minuteStep === 1 ? 'minute' : 'minutes'}`;
  }

  const hourStep = stepOf(hourF);
  if (hourStep && dayF === '*' && monthF === '*' && weekdayF === '*') {
    const minutes = [...parsed.minutes].sort((a, b) => a - b);
    const at = minutes.length === 1 && minutes[0] === 0 ? '' : ` at ${minutes[0]} past`;
    return `every ${hourStep} ${hourStep === 1 ? 'hour' : 'hours'}${at}`;
  }

  const dayStep = stepOf(dayF);
  if (dayStep && monthF === '*' && weekdayF === '*' && parsed.hours.size === 1) {
    const hour = [...parsed.hours][0];
    const minute = Math.min(...parsed.minutes);
    return `every ${dayStep} ${dayStep === 1 ? 'day' : 'days'} at ${clockLabel(hour, minute)}`;
  }

  if (minuteF === '*' && hourF === '*') return 'every minute';

  if (parsed.hours.size !== 1 || parsed.minutes.size !== 1) {
    return `at cron schedule ${parsed.expression}`;
  }

  const at = clockLabel([...parsed.hours][0], [...parsed.minutes][0]);

  let when: string;
  if (weekdayF !== '*') {
    when = weekdayPhrase(weekdayF, parsed.weekdays);
  } else if (dayF !== '*') {
    const days = [...parsed.days]
      .sort((a, b) => a - b)
      .map(ordinal)
      .join(' and ');
    when = `on the ${days}`;
  } else {
    when = 'every day';
  }

  if (monthF !== '*') {
    const months = [...parsed.months]
      .sort((a, b) => a - b)
      .map((m) => MONTH_NAMES[m - 1])
      .join(' and ');
    when = `${when} in ${months}`;
  }

  return `${when} at ${at}`;
}
